instrumentation/mongodb/src/main/java/brave/mongodb/TraceMongoCommandListener.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    ConnectionDescription connectionDescription = event.getConnectionDescription();
+    if (connectionDescription != null) {
+      ConnectionId connectionId = connectionDescription.getConnectionId();
+      if (connectionId != null) {
+        span.tag("mongodb.cluster.id", connectionId.getServerId().getClusterId().getValue());
+      }
+
+      try {
+        InetSocketAddress socketAddress = connectionDescription.getServerAddress().getSocketAddress();
+        span.remoteIpAndPort(socketAddress.getAddress().getHostAddress(), socketAddress.getPort());
+      } catch (MongoSocketException ignored) {
+
+      }
+    }
+
+    span.start();
+  }
+
+  @Override public void commandSucceeded(CommandSucceededEvent event) {
+    Span span = threadLocalSpan.remove();
+    if (span == null) return;
+    span.finish();
+  }
+
+  @Override public void commandFailed(CommandFailedEvent event) {
+    Span span = threadLocalSpan.remove();
+    if (span == null) return;
+    Throwable throwable = event.getThrowable();
+    span.error(throwable == null ? new MongoException("Command failed but no throwable was reported") : throwable);
+    span.finish();
+  }
+
+  @Nullable String getCollectionName(BsonDocument command, String commandName) {
+    if (commandsWithCollectionName.contains(commandName)) {
+      String collectionName = getNonEmptyBsonString(command.get(commandName));
+      if (collectionName != null) {
+        return collectionName;
+      }
+    }
+    // Some other commands, like getMore, have a field like {"collection": collectionName}.
+    return getNonEmptyBsonString(command.get("collection"));
+  }
+
+  /**
+   * @return trimmed string from {@code bsonValue} or null if the trimmed string was empty or the value wasn't a string
+   */
+  @Nullable static String getNonEmptyBsonString(BsonValue bsonValue) {
+    if (bsonValue == null || !bsonValue.isString()) return null;
+    String stringValue = bsonValue.asString().getValue().trim();
+    return stringValue.isEmpty() ? null : stringValue;
+  }
+
+  /**
+   * Returns an abbreviated version of the command for logging/tracing purposes. Currently this is simply a
+   * truncated version of the command's JSON string representation.
+   *
+   * Note that sensitive data related to the MongoDB protocol itself is already scrubbed at this point according to
+   * https://github.com/mongodb/specifications/blob/master/source/command-monitoring/command-monitoring.rst#security
+   *
+   * @return an abbreviated version of the command for logging/tracing purposes
+   */
+  @Nullable String getAbbreviatedCommand(BsonDocument command) {
+    if (maxAbbreviatedCommandLength <= 0) return null;
+    String abbreviatedCommand = command.toJson(jsonWriterSettings);
+    return abbreviatedCommand.isEmpty() ? null : abbreviatedCommand;
+  }
+
+  static String getSpanName(String commandName, @Nullable String collectionName) {
+    return commandName + (collectionName == null ? "" : (" " + collectionName));
```

Comment:
Just return `commandName` if collection is null. I'm not sure but don't expect the JVM to optimize the stringbuilder away for this pattern.